runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -0,0 +1,311 @@
+ * Notice that, unlike the BackpointerStreamView, addresses are discovered
+ * in a single call to the sequencer, therefore, entries are not read
+ * in advance. For this reason, when iterating over the stream (getNextEntry)
+ * we perform batch reads, which significantly reduces the number of RPCs to
+ * Log Units, as opposed to reading one entry at a time. Also, single stepping
+ * in the presence of holes is no longer a need as address space is determined
+ * by the map provided by the sequencer.
+ *
+ * Created by annym on 04/25/19.
+ */
+@Slf4j
+public class AddressMapStreamView extends AbstractQueuedStreamView {
+
+    private static final int COUNTER_INIT_VALUE = 0;
+
+    private static final int DIFF_CONSECUTIVE_ADDRESSES = 1;
+
+    private int batchSize;
+
+    private long addressCount = 0L;
+
+    final StreamOptions options;
+
+    /** Create a new address map stream view.
+     *
+     * @param runtime   The runtime to use for accessing the log.
+     * @param streamId  The ID of the stream to view.
+     */
+    public AddressMapStreamView(final CorfuRuntime runtime,
+                                final UUID streamId,
+                                @Nonnull final StreamOptions options) {
+        super(runtime, streamId);
+        this.options = options;
+        batchSize = runtime.getParameters().getStreamBatchSize();
+    }
+
+    public AddressMapStreamView(final CorfuRuntime runtime,
+                                final UUID streamId) {
+        this(runtime, streamId, StreamOptions.DEFAULT);
+    }
+
+    @Override
+    protected ILogData removeFromQueue(NavigableSet<Long> queue) {
+        // Because the discovery mechanism implemented by this class
+        // does not require to read log entries in advance (requests
+        // full address map in a single call), entries can be read in
+        // batches whenever we have a cache miss. This allows next reads
+        // to be serviced immediately, rather than reading one entry at a time.
+        Long thisRead = queue.pollFirst();
+
+        if (thisRead != null) {
+            // In the case that the immediately requested address (thisRead) is not available,
+            // read it along with a batch of subsequent addresses.
+            List<Long> batchRead = getBatch(thisRead, queue);
+            ILogData ld = readRange(thisRead, batchRead);
+
+            if (queue == getCurrentContext().readQueue) {
+                // Validate data entry belongs to this stream, otherwise, skip.
+                // This verification protects from sequencer regression (tokens assigned in an older epoch
+                // that were never written to, and reassigned on a newer epoch)
+                if (ld.containsStream(this.id)) {
+                    addToResolvedQueue(getCurrentContext(), thisRead, ld);
+                    return ld;
+                } else {
+                    log.trace("getNextEntry[{}]: the data for address {} does not belong to this stream. Skip.",
```

Comment:
Add comment specifying regression.

Reply:
Oh it's actually commented previous to the if statement.